File: sdks/local-rp/zig/src/rpc.zig
//! CSIL-RPC over the injected `Transport`, meant to be TLS-pinned to a
//! domain's DNS `fp=` records — this SDK's only network surface, per the
//! design doc's "Required Network Access": domain public keys, revocations,
//! and claim-ticket redemption, all unauthenticated-TLS TCP CSIL-RPC calls
//! pinned the same way `crates/linkkeys/src/tcp/tls.rs` pins the S2S path.
//!
//! No csilgen Zig target exists (this SDK files that request alongside
//! itself), so the request/response envelope and stream framing here are
//! hand-written directly against `~/repos/catalystcommunity/csilgen/docs/csil-rpc-transport.md`
//! (byte-stream carrier: 4-byte big-endian length prefix + one canonical
//! CBOR envelope; `payload` is CBOR tag-24-wrapped) and
//! `csil-transport-conventions.md` (map keys sorted bytewise — our
//! `cbor.zig` encoder is already canonical by construction).
//!
//! **Pinned TLS is not implemented** — see `tls_pin.zig`'s module docs for
//! the full evaluation of why (`std.crypto.tls.Client` cannot expose the
//! peer certificate after a handshake). `defaultSecureDial` always returns
//! `error.PinnedTlsUnavailable` rather than silently connecting unpinned;
//! callers (including this SDK's own flow tests) inject an alternate
//! `SecureDial` — see this SDK's README.

const std = @import("std");
const cbor = @import("cbor.zig");
const types = @import("types.zig");
const dnsmod = @import("dns.zig");
const transportmod = @import("transport.zig");
const revocation = @import("revocation.zig");

pub const rpc_version: u64 = 1;

/// Mirrors the reference SDKs' own cap, so a malicious/compromised peer
/// cannot drive this client to an unbounded allocation via a forged length
/// prefix.
pub const max_frame_size: usize = 1024 * 1024;

// ---------------------------------------------------------------------
// Envelope
// ---------------------------------------------------------------------

fn untag24(v: cbor.Value) ![]const u8 {
    if (v != .tag or v.tag.number != 24) return error.ExpectedTag24Payload;
    return cbor.asBytes(v.tag.value.*);
}

pub fn encodeRpcRequest(allocator: std.mem.Allocator, service: []const u8, op: []const u8, payload: []const u8) ![]u8 {
    var payload_bytes_val = cbor.bytesVal(payload);
    const entries = [_]cbor.Entry{
        .{ .key = cbor.text("v"), .value = cbor.uint(rpc_version) },
        .{ .key = cbor.text("service"), .value = cbor.text(service) },
        .{ .key = cbor.text("op"), .value = cbor.text(op) },
        .{ .key = cbor.text("payload"), .value = cbor.tagVal(24, &payload_bytes_val) },
    };
    return cbor.encodeAlloc(allocator, cbor.mapVal(&entries));
}

pub const RpcRequestDecoded = struct {
    service: []const u8,
    op: []const u8,
    payload: []const u8,
};

/// Decodes a `CsilRpcRequest` envelope. Only used by this SDK's own test
/// fixtures (fake IDPs) — production code only ever sends requests
/// (`encodeRpcRequest`), never decodes them.
pub fn decodeRpcRequest(allocator: std.mem.Allocator, bytes: []const u8) !RpcRequestDecoded {
    const root = try cbor.decode(allocator, bytes);
    const v = try cbor.asU64(try cbor.require(root, "v"));
    if (v != rpc_version) return error.RpcVersionUnsupported;
    const payload = try untag24(try cbor.require(root, "payload"));
    const service = try cbor.asText(try cbor.require(root, "service"));
    const op = try cbor.asText(try cbor.require(root, "op"));
    return .{ .service = service, .op = op, .payload = payload };
}

/// Encodes a successful (status-Ok) `CsilRpcResponse`. Only used by this
/// SDK's own test fixtures (fake IDPs) — production code only ever decodes
/// responses (`decodeRpcResponse`).
pub fn encodeRpcResponseOk(allocator: std.mem.Allocator, variant: []const u8, payload: []const u8) ![]u8 {
    var payload_val = cbor.bytesVal(payload);
    const entries = [_]cbor.Entry{
        .{ .key = cbor.text("v"), .value = cbor.uint(rpc_version) },
        .{ .key = cbor.text("status"), .value = cbor.uint(0) },
        .{ .key = cbor.text("variant"), .value = cbor.text(variant) },
        .{ .key = cbor.text("payload"), .value = cbor.tagVal(24, &payload_val) },
    };
    return cbor.encodeAlloc(allocator, cbor.mapVal(&entries));
}

/// Encodes a transport-level-failure `CsilRpcResponse` (non-zero status, no
/// typed payload). Only used by this SDK's own test fixtures.
pub fn encodeRpcResponseError(allocator: std.mem.Allocator, status_code: u64, message: []const u8) ![]u8 {
    var empty_payload_val = cbor.bytesVal(&.{});
    const entries = [_]cbor.Entry{
        .{ .key = cbor.text("v"), .value = cbor.uint(rpc_version) },
        .{ .key = cbor.text("status"), .value = cbor.uint(status_code) },
        .{ .key = cbor.text("error"), .value = cbor.text(message) },
        .{ .key = cbor.text("payload"), .value = cbor.tagVal(24, &empty_payload_val) },
    };
    return cbor.encodeAlloc(allocator, cbor.mapVal(&entries));
}

pub const RpcResponse = struct {
    status: i64,
    variant: ?[]const u8 = null,
    error_message: ?[]const u8 = null,
    payload: []const u8 = &.{},
};

pub fn decodeRpcResponse(allocator: std.mem.Allocator, bytes: []const u8) !RpcResponse {
    const root = try cbor.decode(allocator, bytes);
    const v = try cbor.asU64(try cbor.require(root, "v"));
    if (v != rpc_version) return error.RpcVersionUnsupported;

    var payload: []const u8 = &.{};
    if (cbor.mapGet(root, "payload")) |p| payload = try untag24(p);

    const status_val = try cbor.require(root, "status");
    const status: i64 = switch (status_val) {
        .uint => |n| @intCast(n),
        .nint => |n| n,
        else => return error.WrongType,
    };

    return .{
        .status = status,
        .variant = if (cbor.mapGet(root, "variant")) |x| try cbor.asText(x) else null,
        .error_message = if (cbor.mapGet(root, "error")) |x| try cbor.asText(x) else null,
        .payload = payload,
    };
}

fn statusToError(code: i64) anyerror {
    return switch (code) {
        0 => unreachable, // Ok is not an error
        1 => error.RpcMalformedEnvelope,
        2 => error.RpcUnknownServiceOrOp,
        3 => error.RpcUnauthenticated,
        4 => error.RpcForbidden,
        5 => error.RpcVersionUnsupported,
        6 => error.RpcInternal,
        7 => error.RpcUnavailable,
        8 => error.RpcDeadlineExceeded,
        else => error.RpcServerError,
    };
}

// ---------------------------------------------------------------------
// Stream framing (4-byte big-endian length prefix)
// ---------------------------------------------------------------------

pub fn writeLengthPrefixed(stream: std.net.Stream, bytes: []const u8) !void {
    if (bytes.len > max_frame_size) return error.FrameTooLarge;
    var prefix: [4]u8 = undefined;
    std.mem.writeInt(u32, &prefix, @intCast(bytes.len), .big);
    try stream.writeAll(&prefix);
    try stream.writeAll(bytes);
}

/// Reads one length-prefixed frame, enforcing the max-frame guard before
/// allocating. Returns `null` at a clean EOF before any byte of a frame.
pub fn readLengthPrefixed(allocator: std.mem.Allocator, stream: std.net.Stream, max_frame: usize) !?[]u8 {
    var prefix: [4]u8 = undefined;
    const n = try stream.readAll(&prefix);
    if (n == 0) return null;
    if (n != 4) return error.ConnectionClosed;
    const len = std.mem.readInt(u32, &prefix, .big);
    if (len > max_frame) return error.FrameTooLarge;
    const buf = try allocator.alloc(u8, len);
    const got = try stream.readAll(buf);
    if (got != len) return error.ConnectionClosed;
    return buf;
}

// ---------------------------------------------------------------------
// Domain endpoint discovery (DNS fp= pinning + tcp= endpoint)
// ---------------------------------------------------------------------

pub const DomainEndpoint = struct {
    fingerprints: []const []const u8,
    tcp_addr: []const u8,
};

/// Looks up a domain's trust anchor + TCP endpoint over DNS TXT. Fails
/// closed: a missing/unparseable record, or a `_linkkeys` record with no
/// `fp=` entries, or a `_linkkeys_apis` record with no `tcp=` entry, is an
/// error — this SDK never proceeds without a fingerprint set to pin to.
pub fn discoverDomainEndpoint(allocator: std.mem.Allocator, dns: dnsmod.DnsResolver, domain: []const u8) !DomainEndpoint {
    const anchor_name = try dnsmod.linkKeysDnsName(allocator, domain);
    const anchor_txts = try dns.txtLookup(allocator, anchor_name);
    var fingerprints: ?[]const []const u8 = null;
    for (anchor_txts) |txt| {
        const rec = dnsmod.parseLinkKeysTxt(allocator, txt) catch continue;
        if (rec.fingerprints.len > 0) {
            fingerprints = rec.fingerprints;
            break;
        }
    }
    const fps = fingerprints orelse return error.NoTrustAnchorRecord;

    const apis_name = try dnsmod.linkKeysApisDnsName(allocator, domain);
    const apis_txts = try dns.txtLookup(allocator, apis_name);
    var tcp_addr: ?[]const u8 = null;
    for (apis_txts) |txt| {
        const apis = dnsmod.parseLinkKeysApisTxt(allocator, txt) catch continue;
        if (apis.tcp) |t| {
            tcp_addr = t;
            break;
        }
    }
    const addr = tcp_addr orelse return error.NoApisEndpointRecord;

    return .{ .fingerprints = fps, .tcp_addr = addr };
}

// ---------------------------------------------------------------------
// Secure dial seam (see module docs: pinned TLS is not implemented)
// ---------------------------------------------------------------------

pub const SecureDial = *const fn (transport: transportmod.Transport, allocator: std.mem.Allocator, endpoint: DomainEndpoint) anyerror!std.net.Stream;

/// Always fails closed. See this module's docs and `tls_pin.zig` for why:
/// `std.crypto.tls.Client` cannot expose the peer certificate needed for
/// the MANDATORY SPKI pin check, so there is no safe default pinned-TLS
/// implementation to fall back to — connecting *unpinned* would silently
/// defeat the entire trust model (design doc: "an unauthenticated, unpinned
/// fetch would let a LAN MITM substitute domain keys and defeat every
/// downstream signature check").
pub fn defaultSecureDial(transport: transportmod.Transport, allocator: std.mem.Allocator, endpoint: DomainEndpoint) anyerror!std.net.Stream {
    _ = transport;
    _ = allocator;
    _ = endpoint;
    return error.PinnedTlsUnavailable;
}

fn call(allocator: std.mem.Allocator, transport: transportmod.Transport, secure_dial: SecureDial, endpoint: DomainEndpoint, service: []const u8, op: []const u8, payload: []const u8) ![]const u8 {
    const stream = try secure_dial(transport, allocator, endpoint);
    defer stream.close();

    const req_bytes = try encodeRpcRequest(allocator, service, op, payload);
    try writeLengthPrefixed(stream, req_bytes);

    const resp_bytes = try readLengthPrefixed(allocator, stream, max_frame_size) orelse return error.ConnectionClosed;
    const resp = try decodeRpcResponse(allocator, resp_bytes);
    if (resp.status != 0) return statusToError(resp.status);
    return resp.payload;
}

// ---------------------------------------------------------------------
// Domain-key + revocation fetch, ticket redemption
// ---------------------------------------------------------------------

/// Fetches `domain`'s currently-trusted public keys: `DomainKeys/get-domain-keys`
/// over TCP CSIL-RPC, pinned to the domain's DNS `fp=` set, with signing keys
/// pinned directly and encryption keys trusted only via a pinned signing
/// key's vouch (`dns.trustKeys`). Always also fetches
/// `DomainKeys/get-revocations` for the same domain — regardless of what the
/// `get-domain-keys` response's `recent_revocations_available` flag says —
/// and drops any key a quorum-verified sibling revocation certificate
/// targets.
///
/// SEC fix (fail-open -> fail-closed): `recent_revocations_available` is an
/// optional performance hint a well-behaved IDP may use to signal "you don't
/// even need to ask"; a compromised/malicious or merely buggy IDP could
/// otherwise use its absence to suppress this SDK from ever learning about a
/// revocation, which is exactly the scenario revocation exists to guard
/// against — so this SDK never uses it to skip the check. A
/// `get-revocations` RPC error (connection failure, non-Ok status, or
/// response decode failure) is FATAL and propagated: this SDK must fail
/// closed rather than silently proceed with a possibly-stale key set an
/// attacker could have engineered by making the endpoint fail. An empty
/// revocation list is normal success (nothing to apply). An empty trusted
/// result (after applying revocations) is `error.NoTrustedDomainKeys` — fail
/// closed, matching the server's own posture.
pub fn fetchDomainKeys(allocator: std.mem.Allocator, transport: transportmod.Transport, secure_dial: SecureDial, dns: dnsmod.DnsResolver, domain: []const u8) ![]types.DomainPublicKey {
    const endpoint = try discoverDomainEndpoint(allocator, dns, domain);

    const req_payload = try types.encodeEmptyRequest(allocator);
    const resp_bytes = try call(allocator, transport, secure_dial, endpoint, "DomainKeys", "get-domain-keys", req_payload);
    const resp = try types.decodeGetDomainKeysResponse(allocator, resp_bytes);

    var trusted = try dnsmod.trustKeys(allocator, resp.keys, endpoint.fingerprints);
    if (trusted.len == 0) return error.NoTrustedDomainKeys;

    // Always fetch revocations — never gated on `recent_revocations_available`
    // (see this function's doc comment). Any failure here propagates via
    // `try`, i.e. is FATAL: it must never be swallowed to "just proceed
    // unfiltered".
    var since_buf: [32]u8 = undefined;
    const local_rp = @import("local_rp.zig");
    const since = try local_rp.formatTimestamp(&since_buf, std.time.timestamp() - 30 * 24 * 60 * 60);
    const rev_req_payload = try types.encodeGetRevocationsRequest(allocator, since);
    const rev_resp_bytes = try call(allocator, transport, secure_dial, endpoint, "DomainKeys", "get-revocations", rev_req_payload);
    const rev_resp = try types.decodeGetRevocationsResponse(allocator, rev_resp_bytes);
    for (rev_resp.revocations) |cert| {
        if (revocation.verifyRevocationCertificate(allocator, cert, trusted, domain)) |_| {
            trusted = removeKeyById(trusted, cert.target_key_id);
        } else |_| {}
    }

    if (trusted.len == 0) return error.NoTrustedDomainKeys;
    return trusted;
}

fn removeKeyById(keys: []types.DomainPublicKey, key_id: []const u8) []types.DomainPublicKey {
    var write: usize = 0;
    for (keys) |k| {
        if (!std.mem.eql(u8, k.key_id, key_id)) {
            keys[write] = k;
            write += 1;
        }
    }
    return keys[0..write];
}

/// Redeems a claim ticket with `domain`'s IDP: `LocalRp/redeem-claim-ticket`
/// over TCP CSIL-RPC, pinned via the domain's DNS `fp=` set. Unauthenticated
/// at the transport layer (no client cert) — the redemption request itself
/// is signed with the local RP's signing key, which is the possession proof
/// the server checks.
pub fn redeemClaimTicket(allocator: std.mem.Allocator, transport: transportmod.Transport, secure_dial: SecureDial, dns: dnsmod.DnsResolver, domain: []const u8, signed_request: types.SignedLocalRpTicketRedemptionRequest) !types.LocalRpTicketRedemptionResponse {
    const endpoint = try discoverDomainEndpoint(allocator, dns, domain);
    const payload = try types.encodeSignedLocalRpTicketRedemptionRequest(allocator, signed_request);
    const resp_bytes = try call(allocator, transport, secure_dial, endpoint, "LocalRp", "redeem-claim-ticket", payload);
    return types.decodeLocalRpTicketRedemptionResponse(allocator, resp_bytes);
}

test "RpcRequest/RpcResponse envelope round trip" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const a = arena.allocator();

    const req_bytes = try encodeRpcRequest(a, "DomainKeys", "get-domain-keys", "hello-payload");
    const decoded_root = try cbor.decode(a, req_bytes);
    try std.testing.expectEqualStrings("DomainKeys", try cbor.asText(try cbor.require(decoded_root, "service")));
    try std.testing.expectEqualStrings("get-domain-keys", try cbor.asText(try cbor.require(decoded_root, "op")));
    const payload_val = try cbor.require(decoded_root, "payload");
    try std.testing.expectEqualStrings("hello-payload", try untag24(payload_val));

    // Build a response the way a server would and decode it back.
    var payload_bytes_val = cbor.bytesVal("resp-payload");
    const resp_entries = [_]cbor.Entry{
        .{ .key = cbor.text("v"), .value = cbor.uint(1) },
        .{ .key = cbor.text("status"), .value = cbor.uint(0) },
        .{ .key = cbor.text("variant"), .value = cbor.text("GetDomainKeysResponse") },
        .{ .key = cbor.text("payload"), .value = cbor.tagVal(24, &payload_bytes_val) },
    };
    const resp_bytes = try cbor.encodeAlloc(a, cbor.mapVal(&resp_entries));
    const resp = try decodeRpcResponse(a, resp_bytes);
    try std.testing.expectEqual(@as(i64, 0), resp.status);
    try std.testing.expectEqualStrings("GetDomainKeysResponse", resp.variant.?);
    try std.testing.expectEqualStrings("resp-payload", resp.payload);
}

test "length-prefix framing is byte-exact" {
    var buf: [4]u8 = undefined;
    std.mem.writeInt(u32, &buf, 5, .big);
    try std.testing.expectEqualSlices(u8, &[_]u8{ 0, 0, 0, 5 }, &buf);
}

test "defaultSecureDial always fails closed" {
    var st = transportmod.StdTransport{};
    const endpoint = DomainEndpoint{ .fingerprints = &.{}, .tcp_addr = "127.0.0.1:1" };
    try std.testing.expectError(error.PinnedTlsUnavailable, defaultSecureDial(st.transport(), std.testing.allocator, endpoint));
}

File: sdks/local-rp/php/src/Rpc.php
<?php

declare(strict_types=1);

namespace LinkKeys\LocalRp;

use Csilgen\Generated\DomainPublicKey;
use Csilgen\Generated\EmptyRequest;
use Csilgen\Generated\GetRevocationsRequest;
use Csilgen\Generated\LocalRpTicketRedemptionResponse;
use Csilgen\Generated\SignedLocalRpTicketRedemptionRequest;

/**
 * CSIL-RPC over the injected {@see Transport}, TLS-pinned to a domain's DNS
 * `fp=` records — this SDK's only network surface (design doc, "Required
 * Network Access"): domain public keys, revocations, and claim-ticket
 * redemption, all unauthenticated-TLS TCP CSIL-RPC calls pinned the same way
 * `crates/linkkeys/src/tcp/tls.rs` pins the S2S path.
 *
 * This class hand-builds the CSIL-RPC envelope
 * (`~/repos/catalystcommunity/csilgen/docs/csil-rpc-transport.md` §1.1/1.2:
 * `{v, service, op, ?id, payload, ?auth}` request / `{v, ?id, status,
 * ?variant, ?error, payload}` response, verbatim CSIL `service`/`op` names,
 * length-prefixed stream framing) rather than driving it through
 * `src/Generated/`'s generated client wrapper classes, for the same reason
 * the Rust and Python reference SDKs do: the generated client passes a
 * transport-agnostic `(service, method)` pair to an injected `Transport`
 * seam that is lossy for reconstructing the real wire names (confirmed
 * against `php-client`'s generated `client.php`: it collapses `service`/`op`
 * into one already-mangled route string, e.g. `'domain-keys/get-domain-keys'`
 * — see the csilgen defect filed for this at
 * `~/repos/catalystcommunity/csilgen/docs/csilgen-requests/`). This SDK only
 * ever calls two operations, so it builds the two real CSIL-RPC requests
 * directly, mirroring `sdks/local-rp/rust/src/rpc.rs` /
 * `sdks/local-rp/python/linkkeys_local_rp/rpc.py`.
 */
final class Rpc
{
    /**
     * Mirrors the server's own cap (`crates/linkkeys-rpc-client/src/lib.rs`)
     * so a malicious/compromised peer cannot drive this client to an
     * unbounded allocation via a forged length prefix.
     */
    public const MAX_FRAME_SIZE = 1024 * 1024;

    private const CSIL_RPC_VERSION = 1;
    private const TAG_ENCODED_CBOR = 24;

    /** @param resource $stream */
    private static function sendFrame($stream, string $data): void
    {
        $len = strlen($data);
        $header = pack('N', $len);
        self::writeAll($stream, $header . $data);
    }

    /** @param resource $stream */
    private static function writeAll($stream, string $data): void
    {
        $total = strlen($data);
        $written = 0;
        while ($written < $total) {
            $n = fwrite($stream, substr($data, $written));
            if ($n === false || $n === 0) {
                $meta = stream_get_meta_data($stream);
                if (!empty($meta['timed_out'])) {
                    throw new RpcTransportError('write timed out');
                }
                throw new RpcTransportError('connection closed while writing');
            }
            $written += $n;
        }
    }

    /** @param resource $stream */
    private static function readExact($stream, int $n): string
    {
        $buf = '';
        while (strlen($buf) < $n) {
            $chunk = fread($stream, $n - strlen($buf));
            if ($chunk === false || $chunk === '') {
                $meta = stream_get_meta_data($stream);
                if (!empty($meta['timed_out'])) {
                    throw new RpcTransportError('read timed out');
                }
                throw new RpcProtocolError('connection closed before expected bytes were received');
            }
            $buf .= $chunk;
        }
        return $buf;
    }

    /** @param resource $stream */
    private static function recvFrame($stream): string
    {
        $lenBytes = self::readExact($stream, 4);
        $len = unpack('N', $lenBytes)[1];
        if ($len > self::MAX_FRAME_SIZE) {
            throw new RpcProtocolError("peer frame too large ({$len} bytes, max " . self::MAX_FRAME_SIZE . ')');
        }
        return self::readExact($stream, $len);
    }

    private static function encodeRequest(string $service, string $op, string $payload): string
    {
        return Cbor::encode([
            'v' => self::CSIL_RPC_VERSION,
            'service' => $service,
            'op' => $op,
            'payload' => Cbor::tag(self::TAG_ENCODED_CBOR, Cbor::bytes($payload)),
        ]);
    }

    /** @return array{status:int, variant:?string, error:?string, payload:string} */
    private static function decodeResponse(string $bytes): array
    {
        $m = Cbor::decode($bytes);
        if (!is_array($m) || !isset($m['status']) || !is_int($m['status'])) {
            throw new RpcProtocolError("RPC response envelope missing integer 'status'");
        }
        $payloadTag = $m['payload'] ?? null;
        $payload = '';
        if ($payloadTag instanceof CborTag && $payloadTag->tag === self::TAG_ENCODED_CBOR) {
            $payload = $payloadTag->value instanceof CborBytes ? $payloadTag->value->data : (string) $payloadTag->value;
        }
        return [
            'status' => $m['status'],
            'variant' => $m['variant'] ?? null,
            'error' => $m['error'] ?? null,
            'payload' => $payload,
        ];
    }

    // -------------------------------------------------------------------
    // Endpoint discovery
    // -------------------------------------------------------------------

    /**
     * Look up a domain's trust anchor + TCP endpoint over DNS TXT. Fails
     * closed: a missing/unparseable record, or a `_linkkeys` record with no
     * `fp=` entries, or a `_linkkeys_apis` record with no `tcp=` entry, is
     * an error — this SDK never proceeds without a fingerprint set to pin
     * to.
     */
    public static function discoverDomainEndpoint(DnsResolver $dns, string $domain): DomainEndpoint
    {
        $anchorName = Dns::linkkeysDnsName($domain);
        $fingerprints = [];
        foreach ($dns->txtLookup($anchorName) as $txt) {
            try {
                $fps = Dns::parseLinkKeysTxt($txt);
            } catch (DnsParseError $e) {
                continue;
            }
            if (!empty($fps)) {
                $fingerprints = $fps;
                break;
            }
        }
        if (empty($fingerprints)) {
            throw new DnsParseError(DnsParseError::NO_LINKKEYS_RECORD, "no usable {$anchorName} TXT record with fp= entries");
        }

        $apisName = Dns::linkkeysApisDnsName($domain);
        $tcpAddr = null;
        foreach ($dns->txtLookup($apisName) as $txt) {
            try {
                $apis = Dns::parseLinkKeysApisTxt($txt);
            } catch (DnsParseError $e) {
                continue;
            }
            if ($apis['tcp'] !== null) {
                $tcpAddr = $apis['tcp'];
                break;
            }
        }
        if ($tcpAddr === null) {
            throw new DnsParseError(DnsParseError::MISSING_APIS_ENDPOINT, "no usable {$apisName} TXT record with tcp= entry");
        }

        return new DomainEndpoint($fingerprints, $tcpAddr);
    }

    /**
     * Open a TLS connection to `$endpoint`, pinned to its fingerprints, over
     * a raw stream obtained from `$transport`. If `$transport` implements
     * {@see OpaqueTransport}, the real TLS-pinning wrap step
     * ({@see Tls::dialTlsPinned}) is skipped and `$transport`'s raw stream
     * is returned as-is — see that interface's docblock for why this exists
     * and why it is safe (no runtime-settable, process-global pin-bypass
     * switch: opting out of pinning requires supplying and injecting an
     * entire `Transport` implementation, a much larger and more visible
     * commitment than flipping a static property).
     *
     * @return resource
     */
    private static function dialTls(Transport $transport, DomainEndpoint $endpoint)
    {
        $raw = $transport->dial($endpoint->tcpAddr);
        if ($transport instanceof OpaqueTransport) {
            return $raw;
        }
        $hostname = Tls::extractHostname($endpoint->tcpAddr);
        try {
            return Tls::dialTlsPinned($raw, $hostname, $endpoint->fingerprints);
        } catch (\Throwable $e) {
            fclose($raw);
            throw $e;
        }
    }

    /** Send one CSIL-RPC request over a fresh TLS connection and return the decoded success payload. */
    private static function call(Transport $transport, DomainEndpoint $endpoint, string $service, string $op, string $payload): string
    {
        $stream = self::dialTls($transport, $endpoint);
        try {
            self::sendFrame($stream, self::encodeRequest($service, $op, $payload));
            $respBytes = self::recvFrame($stream);
        } finally {
            fclose($stream);
        }

        $resp = self::decodeResponse($respBytes);
        if ($resp['status'] !== 0) {
            throw new RpcServerError($resp['status'], $resp['error'] ?? 'unknown error');
        }
        return $resp['payload'];
    }

    // -------------------------------------------------------------------
    // High-level operations
    // -------------------------------------------------------------------

    /**
     * Fetch `$domain`'s currently-trusted public keys:
     * `DomainKeys/get-domain-keys` over TCP CSIL-RPC, pinned to the domain's
     * DNS `fp=` set, with signing keys pinned directly and encryption keys
     * trusted only via a pinned signing key's vouch ({@see Dns::trustKeys}).
     * ALWAYS also fetches `DomainKeys/get-revocations` for the same domain —
     * this is not suppressible by `recent_revocations_available` or any
     * other server-advertised hint (design doc, "Post-implementation
     * security review", item 4: a server-controlled flag must never be able
     * to suppress revocation delivery) — and drops any key a
     * quorum-verified sibling revocation certificate targets. A
     * `get-revocations` RPC/decode error is FATAL: this method fails closed
     * rather than verifying the callback against a possibly-stale,
     * unfiltered key set. An empty revocations list is a normal success (no
     * keys are revoked). An empty trusted result throws
     * {@see NoTrustedDomainKeysError} — fail closed.
     *
     * @return DomainPublicKey[]
     */
    public static function fetchDomainKeys(Transport $transport, DnsResolver $dns, string $domain, \DateTimeImmutable $now): array
    {
        $endpoint = self::discoverDomainEndpoint($dns, $domain);

        $payload = Wire::encodeEmptyRequest(new EmptyRequest());
        $respBytes = self::call($transport, $endpoint, 'DomainKeys', 'get-domain-keys', $payload);
        $resp = Wire::decodeGetDomainKeysResponse($respBytes);

        $trusted = Dns::trustKeys($resp->keys, $endpoint->fingerprints, $now);
        if (empty($trusted)) {
            throw new NoTrustedDomainKeysError($domain);
        }

        $since = Time::toRfc3339($now->sub(new \DateInterval('P30D')));
        $reqPayload = Wire::encodeGetRevocationsRequest(new GetRevocationsRequest(['since' => $since]));
        try {
            $respBytes = self::call($transport, $endpoint, 'DomainKeys', 'get-revocations', $reqPayload);
            $revResp = Wire::decodeGetRevocationsResponse($respBytes);
        } catch (\Throwable $e) {
            // Fail closed: we cannot verify the callback against a key set
            // we don't know is free of recently-revoked keys.
            throw new RevocationFetchError($domain, $e);
        }
        foreach ($revResp->revocations as $cert) {
            try {
                Revocation::verifyRevocationCertificate($cert, $trusted, $domain);
                $trusted = array_values(array_filter($trusted, fn ($k) => $k->keyId !== $cert->targetKeyId));
            } catch (RevocationError $e) {
                // Quorum not met for this particular certificate; ignore
                // just that certificate, not the whole revocation fetch.
            }
        }

        if (empty($trusted)) {
            throw new NoTrustedDomainKeysError($domain);
        }
        return $trusted;
    }

    /**
     * Redeem a claim ticket with `$domain`'s IDP:
     * `LocalRp/redeem-claim-ticket` over TCP CSIL-RPC, pinned via the
     * domain's DNS `fp=` set. Unauthenticated at the transport layer (no
     * client cert) — the redemption request itself is signed with the local
     * RP's signing key, which is the possession proof the server checks.
     */
    public static function redeemClaimTicket(Transport $transport, DnsResolver $dns, string $domain, SignedLocalRpTicketRedemptionRequest $signedRequest): LocalRpTicketRedemptionResponse
    {
        $endpoint = self::discoverDomainEndpoint($dns, $domain);
        $payload = Wire::encodeSignedLocalRpTicketRedemptionRequest($signedRequest);
        $respBytes = self::call($transport, $endpoint, 'LocalRp', 'redeem-claim-ticket', $payload);
        return Wire::decodeLocalRpTicketRedemptionResponse($respBytes);
    }
}

/** Discovered endpoint for a domain: its pinned trust-anchor fingerprints (`_linkkeys`) and its CSIL-RPC TCP address (`_linkkeys_apis` `tcp=`). */
final class DomainEndpoint
{
    /** @var string[] */
    public array $fingerprints;
    public string $tcpAddr;

    /** @param string[] $fingerprints */
    public function __construct(array $fingerprints, string $tcpAddr)
    {
        $this->fingerprints = $fingerprints;
        $this->tcpAddr = $tcpAddr;
    }
}

class RpcError extends \RuntimeException
{
}

final class RpcProtocolError extends RpcError
{
}

final class RpcTransportError extends RpcError
{
}

final class RpcServerError extends RpcError
{
    public int $status;
    public string $errorMessage;

    public function __construct(int $status, string $message)
    {
        parent::__construct("server error ({$status}): {$message}");
        $this->status = $status;
        $this->errorMessage = $message;
    }
}

final class NoTrustedDomainKeysError extends RpcError
{
    public string $domain;

    public function __construct(string $domain)
    {
        parent::__construct("no trusted public keys could be established for domain: {$domain}");
        $this->domain = $domain;
    }
}

/**
 * `DomainKeys/get-revocations` failed (transport, protocol, or decode
 * error). Fetching revocations is unconditional and this error is fatal —
 * {@see Rpc::fetchDomainKeys} fails closed rather than trusting a key set it
 * couldn't check for recent revocations against.
 */
final class RevocationFetchError extends RpcError
{
    public string $domain;

    public function __construct(string $domain, ?\Throwable $previous = null)
    {
        parent::__construct("failed to fetch revocations for domain: {$domain}", 0, $previous);
        $this->domain = $domain;
    }
}
